core/src/avm2/bytearray.rs

```diff
@@ -33,316 +34,210 @@ impl ByteArrayStorage {
-            self.bytes.resize(offset, 0);
+    /// Reads any amount of bytes from the current position in the ByteArray
+    #[inline]
+    pub fn read_bytes(&self, amnt: usize) -> Result<&[u8], Error> {
+        let bytes = self.read_at(amnt, self.position.get())?;
+        self.position.set(self.position.get() + amnt);
+        Ok(bytes)
+    }
+
+    /// Reads any amount of bytes at any offset in the ByteArray
+    #[inline]
+    pub fn read_at(&self, amnt: usize, offset: usize) -> Result<&[u8], Error> {
+        self.bytes
+            .get(
+                offset
+                    ..offset
+                        .checked_add(amnt)
+                        .ok_or("RangeError: Cannot overflow usize")?,
+            )
+            .ok_or_else(|| "EOFError: Reached EOF".into())
+    }
+
+    /// Write bytes at any offset in the ByteArray
+    /// Will automatically grow the ByteArray to fit the new buffer
+    pub fn write_at(&mut self, buf: &[u8], offset: usize) -> Result<(), Error> {
+        let new_len = offset
+            .checked_add(buf.len())
+            .ok_or("RangeError: The length of this ByteArray is too big")?;
+        if self.len() < new_len {
+            self.resize(new_len, 0);
         }
-        // Figure out what bytes will be used to overwrite what's currently
-        // there (left), and what will be appended on the end (right)
-        {
-            let space = self.bytes.len() - offset;
-            let (left, right) = buf.split_at(cmp::min(space, buf.len()));
-            self.bytes[offset..offset + left.len()].copy_from_slice(left);
-            self.bytes.extend_from_slice(right);
-        }
-    }
-
-    pub fn clear(&mut self) {
-        self.bytes.clear();
-        // According to docs, this is where the bytearray should free resources
-        self.bytes.shrink_to_fit();
-        self.position = 0;
-    }
-
-    // Returns the bytearray compressed with zlib
-    pub fn zlib_compress(&mut self) -> io::Result<Vec<u8>> {
-        let mut buffer = Vec::new();
-        let mut compresser = ZlibEncoder::new(&*self.bytes, Compression::fast());
-        compresser.read_to_end(&mut buffer)?;
-        Ok(buffer)
+        // SAFETY:
+        // The storage is garunteed to be at least the size of new_len because we just resized it.
+        unsafe { self.get_unchecked_mut(offset..new_len).copy_from_slice(buf) }
+        Ok(())
     }
 
-    // Returns the bytearray compressed with deflate
-    pub fn deflate_compress(&mut self) -> io::Result<Vec<u8>> {
-        let mut buffer = Vec::new();
-        let mut compresser = DeflateEncoder::new(&*self.bytes, Compression::fast());
-        compresser.read_to_end(&mut buffer)?;
-        Ok(buffer)
+    /// Write bytes at any offset in the ByteArray
+    /// Will return an error if the new buffer does not fit the ByteArray
+    pub fn write_at_nongrowing(&mut self, buf: &[u8], offset: usize) -> Result<(), Error> {
```

Comment:
I might prefer `copy_from` or `copy_from_slice` here instead of `write_at_nongrowing`. That way we can differentiate the `write` methods, which grow the storage, from `copy` which will not grow, but are fallible.